### ISSUE TITLE
Class 'BaseFilter<T, P, M>' incorrectly implements interface 'IFilterComp'.

### ISSUE THREAD

**Issue by katasup** [ag-grid/ag-grid#1889]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

```
[at-loader] Using typescript@2.4.2 from typescript and "tsconfig.json" from /Users/username/projects/myProject/tsconfig.json.

[at-loader] Checking started in a separate process...

[at-loader] Checking finished with 1 errors
Time: 41624ms
chunk    {0} app.min.js, app.css (app) 484 kB {1} [initial] [rendered]
chunk    {1} vendor.js, vendor.css (vendor) 4.11 MB [entry] [rendered]

ERROR in [at-loader] ./node_modules/ag-grid/dist/lib/filter/baseFilter.d.ts:21:31 
    TS2420: Class 'BaseFilter<T, P, M>' incorrectly implements interface 'IFilterComp'.
  Types of property 'afterGuiAttached' are incompatible.
    Type '(params: IAfterGuiAttachedParams) => void' is not assignable to type '((params?: IFilterAfterGuiAttachedParams | undefined) => void) | undefined'.
      Type '(params: IAfterGuiAttachedParams) => void' is not assignable to type '(params?: IFilterAfterGuiAttachedParams | undefined) => void'.
        Types of parameters 'params' and 'params' are incompatible.
          Type 'IFilterAfterGuiAttachedParams | undefined' is not assignable to type 'IAfterGuiAttachedParams'.
            Type 'undefined' is not assignable to type 'IAfterGuiAttachedParams'.
```

**Expected behavior**

**Minimal reproduction of the problem with instructions**

webpack@3.5.4, awesome-typescript-loader@3.2.2

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

MacOS Sierra, VS Code, Node.js 8.4.0

* **ag-Grid version:** 13.1.0

* **Browser:**

* **Language:** TypeScript@2.4.2

**Comment by makinggoodsoftware** [comment]
Hi,

This seems to be compiling fine in our end, can you check if you are not using mixed versions of ag-grid in your modules (ie all are 13.1.0)

Thanks

**Comment by katasup** [comment]
This is only packages found by `/ag-grid/g` in my package.json:
```
    "ag-grid": "^13.1.0",
    "ag-grid-enterprise": "^13.1.0",
    "ag-grid-react": "^13.1.0",
```
Same for package-lock.json:
```
    "ag-grid": {
      "version": "13.1.0",
      "resolved": "https://registry.npmjs.org/ag-grid/-/ag-grid-13.1.0.tgz",
      "integrity": "sha1-CQYnEU5oblStZcyzFvh6ZyZjwFM="
    },
    "ag-grid-enterprise": {
      "version": "13.1.0",
      "resolved": "https://registry.npmjs.org/ag-grid-enterprise/-/ag-grid-enterprise-13.1.0.tgz",
      "integrity": "sha1-b/rr8hawPoCjrQD+VeZDEHQ0uRA="
    },
    "ag-grid-react": {
      "version": "13.1.0",
      "resolved": "https://registry.npmjs.org/ag-grid-react/-/ag-grid-react-13.1.0.tgz",
      "integrity": "sha1-blOTpTgEK/K+JzoUWyfEP1vZLZc="
    },
```

**Comment by katasup** [comment]
Okay, i have a solution: set "strictNullChecks" at tsconfig.json to true and get error.
Can do u fix this?

**Comment by manu-st** [comment]
I also have the issue as we are using strictNullChecks.

**Comment by RapidTransit** [comment]
Also having this issue :(, The work around fixes it

**Comment by UselessPickles** [comment]
Something worth considering (requires someone knowledgable of this codebase, intent, etc)...

There are 2 "solutions" that fix the compiler error:

1. Update the signature of IComponent.afterGuiAttached to make "params" NON-optional.
2. Update the signature of the method Component.agterGuiAttached to make "params" OPTIONAL, and update the implementation to handle the fact that "params" may be undefined.

Which is actually the correct solution? Is "params" truely optional? Or will a non-undefined/null value always be passed to the method?

**Comment by UselessPickles** [comment]
Can this issue get re-opened/fixed? As others have pointed out, this error occurs when strict null checks are enabled in the TypeScript compiler. I would like to use strict null checks in my project, and think it is unfair to be forced to disable the benefits of that feature in order to pass compilation with ag-grid.

**Comment by mast4461** [comment]
It seems that this problem is fixed since at least [v14.1.1](https://github.com/ag-grid/ag-grid/releases/tag/14.1.1). 😃
